### PR TITLE
net: coap_client: Fix non-confirmable requests

### PR DESCRIPTION
**Pull request by SeppoTakalo**
This PR contains few minor refactoring of test cases and a bug fix for Non-Confirmable requests

Non-confirmable CoAP requests were just cumulating in the client structure, if there was no response. This was noticed by @jeffwelder-ellenbytech  in https://github.com/zephyrproject-rtos/zephyr/pull/81282#issuecomment-2479619306

Minor changes:
* Fix timeout when waiting for separate response
* Refactored test cases to be easier to copy&paste
* Refactored tests to use realt-time scheduler, so it is more accurate
* Use const pointer in coap_requests

### CODE REVIEW

**Comment by SeppoTakalo** [comment]
@jeffwelder-ellenbytech  please check.

You found a real bug in the non-confirmable handling.

File: subsys/net/lib/coap/coap_client.c [hunk 1]
```diff
@@ -521,6 +533,11 @@ static void coap_client_resend_handler(struct coap_client *client)
 
 	for (int i = 0; i < CONFIG_COAP_CLIENT_MAX_REQUESTS; i++) {
 		if (timeout_expired(&client->requests[i])) {
+			if (!client->requests[i].coap_request.confirmable) {
+				release_internal_request(&client->requests[i]);
```
Comment (jeffwelder-ellenbytech) [inline subsys/net/lib/coap/coap_client.c:537]: Perhaps we can still belly up to application that this request has been internally released due to no-response, so the appropriate cleanup can happen.

`report_callback_error(&client->requests[i], -ENOMSG);`

**Comment by jeffwelder-ellenbytech** [comment]
Awesome, was just looking at this area again this morning. :)

I think you covered the case elegantly, it seems to be working through my tests.

If you send a non-confirmable request from the application, and there is no response, the request gets released silently without calling the callback handler (making the user have to duplicate the timeout mechanism for their allocated response handler object/memory). Maybe this could be a response for `-ENOMSG` or something more appropriate? [See comment on PR]

With a confirmable message the handler still gets called with a `-ETIMEDOUT` error after n retries, giving an explicit place to free the application side's request.

**Comment by SeppoTakalo** [comment]
> without calling the callback handler (making the user have to duplicate the timeout mechanism for their allocated response handler object/memory). Maybe this could be a response for `-ENOMSG` or something more appropriate?

What would you do in the application as a result of that callback?

To be honest, I don't really understand the point of non-confirmable messages. Most use cases I have seen, tend to use more code or even more messages to cope with the fact that those are not retried, and delivery cannot be trusted.

So on most use cases, where someone thinks about having non-confirmable, but still have a callback, the proper answer would probably be to use confirmable. I don't really understand why to use non-confirmable but still expect an answer back. Therefore I am a bit against of delivering an error code as a result of perfectly functioning system where server does not send anything as a response to NON message.

Technically, you could just

```c
/* No callback, just send and forget */
static struct coap_client_request req = {
	.method = COAP_METHOD_POST,
	.confirmable = false,
	.path = "/hello",
};
coap_client_req(&client, client.fd, 0, &req, NULL);
zsock_close(client.fd); /* doesn't matter, message already went out (or got dropped) */
```

**Review by rlubos — APPROVED** [review]
Looks good.

**Comment by jeffwelder-ellenbytech** [comment]
Agreed, non-confirmable is a strange animal.

I think you covered our use case of pings/network statistics where we don't respond on the server, and don't want retries to save on cellular data.

Oddly the [(CoAP RFC) 5.2.3. Non-confirmable Section](https://datatracker.ietf.org/doc/html/rfc7252#section-5.2.3) does say that a confirmable message can come back from the server in response to a non-confirmable request. If that happens, having this callback is the only way to know when to cleanup the request in the application.

> What would you do in the application as a result of that callback?

Our application has an array of request structs because we allow concurrent requests. If a data response is possible we need to maintain that array. (Right now we just drop non-confirmable requests because we don't expect a response)

> Therefore I am a bit against of delivering an error code as a result of perfectly functioning system where server does not send anything as a response to NON message.

I agree, but I think it's covered if you don't set a handler for the request, you wont get this error code raised.

This isn't really a benefit for our use case, and I agree with your point about being the wrong usage of non-confirmable, but if someone wanted to use coap_client for this or had less control over the server, it would give them a hook. 🤷‍♂️

**Review by jukkar — APPROVED** [review]

**Pull request merged**